src/main.rs

```diff
@@ -1,41 +1,83 @@
-use std::io::{stdin, stdout, stderr, Read, Write};
-use std::io;
+extern crate memchr;
+
+use std::io::{self, Read, Write};
 use std::process::exit;
+use memchr::memchr;
 
-fn has_data(size: usize) -> Option<usize> {
-    match size {
-        0 => None,
-        x => Some(x),
-    }
+struct NewlineWrapper<R: Read> {
+    inner: R
 }
 
-fn run() -> io::Result<()> {
-    let mut stdin = stdin();
-    let mut stdout = stdout();
-    let mut buf = [0; 4096];
-
-    while let Some(size) = stdin.read(&mut buf).map(has_data)? {
-        let mut flush = false;
-
-        for b in &mut buf[0..size] {
-            if *b == b'\r' {
-                *b = b'\n';
-                flush = true;
-            }
+impl<R: Read> NewlineWrapper<R> {
+    fn new(inner: R) -> Self {
+        NewlineWrapper {
+            inner: inner,
         }
+    }
+}
 
-        stdout.write(&buf[0..size])?;
-        if flush {
-            stdout.flush()?;
+impl<R: Read> Read for NewlineWrapper<R> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        let size = self.inner.read(buf)?;
+        let mut start = 0;
+        while start < size {
+            let to_check = &mut buf[start..size];
+            match memchr(b'\r', to_check) {
+                Some(i) => {
+                    to_check[i] = b'\n';
+                    start += i + 1;
+                },
+                None => break,
+            }
         }
+        Ok(size)
     }
+}
+
+fn run() -> io::Result<()> {
+    let stdin = io::stdin();
+    let stdin = stdin.lock();
+    let mut stdin = NewlineWrapper::new(stdin);
+    let stdout = io::stdout();
+    let stdout = stdout.lock();
+    let mut stdout = io::LineWriter::new(stdout);
```

Comment:
I don't think that LineWriter is more performant than my previous implementation - I expect that it searches memory for '`\n` second time. That is unnecessary.

Of course, my implementation doesn't flush on `\n` in input stream (only on `\r`). This was intentional. However, I can imagine having two modes (changeable by switch).